java/source/data.rst

```diff
@@ -0,0 +1,316 @@
+
+   import org.apache.arrow.vector.IntVector;
+   import org.apache.arrow.vector.compare.TypeEqualsVisitor;
+
+   IntVector right = new IntVector("int", rootAllocator);
+   IntVector left1 = new IntVector("int", rootAllocator);
+   IntVector left2 = new IntVector("int2", rootAllocator);
+
+   setVector(right, 10,20,30);
+
+   TypeEqualsVisitor visitor = new TypeEqualsVisitor(right); // equal or unequal
+
+Comparing vector fields:
+
+.. code-block:: java
+   :emphasize-lines: 1-4
+
+   jshell> visitor.equals(left1); visitor.equals(left2);
+
+   true
+   false
+
+Compare values on the array
+===========================
+
+.. code-block:: java
+   :emphasize-lines: 15-17
+
+   import org.apache.arrow.algorithm.sort.StableVectorComparator;
+   import org.apache.arrow.algorithm.sort.VectorValueComparator;
+   import org.apache.arrow.vector.VarCharVector;
+
+   // compare two values at the given indices in the vectors.
+   // comparing org.apache.arrow.algorithm.sort.VectorValueComparator on algorithm
+   VarCharVector vec = new VarCharVector("valueindexcomparator", rootAllocator);
+   vec.allocateNew(100, 5);
+   vec.setValueCount(10);
+   vec.set(0, "ba".getBytes());
+   vec.set(1, "abc".getBytes());
+   vec.set(2, "aa".getBytes());
+   vec.set(3, "abc".getBytes());
+   vec.set(4, "a".getBytes());
+   VectorValueComparator<VarCharVector> comparatorValues = new TestVarCharSorter(); // less than, equal to, greater than
+   VectorValueComparator<VarCharVector> stableComparator = new StableVectorComparator<>(comparatorValues);//Stable comparator only supports comparing values from the same vector
+   stableComparator.attachVector(vec);
+
+Comparing two values at the given indices in the vectors:
+
+.. code-block:: java
+   :emphasize-lines: 1-8
+
+   jshell> stableComparator.compare(0, 1) > 0; stableComparator.compare(1, 2) < 0; stableComparator.compare(2, 3) < 0; stableComparator.compare(1, 3) < 0; stableComparator.compare(3, 1) > 0; stableComparator.compare(3, 3) == 0;
+
+   true
+   true
+   true
+   true
+   true
+   true
+
+Search values on the array
+==========================
+
+Linear search - O(n)
+********************
+
+Algorithm: org.apache.arrow.algorithm.search.VectorSearcher#linearSearch - O(n)
+
+.. code-block:: java
+   :emphasize-lines: 27
+
+   import org.apache.arrow.algorithm.search.VectorSearcher;
+   import org.apache.arrow.algorithm.sort.DefaultVectorComparators;
+   import org.apache.arrow.algorithm.sort.VectorValueComparator;
+   import org.apache.arrow.vector.IntVector;
+
+   // search values on the array
+   // linear search org.apache.arrow.algorithm.search.VectorSearcher#linearSearch - O(n)
+   IntVector rawVector = new IntVector("", rootAllocator);
+   IntVector negVector = new IntVector("", rootAllocator);
+   rawVector.allocateNew(10);
+   rawVector.setValueCount(10);
+   negVector.allocateNew(1);
+   negVector.setValueCount(1);
+   for (int i = 0; i < 10; i++) { // prepare data in sorted order
+    if (i == 0) {
```

Comment:
Can we try to be consistent about indent spacing?

Reply:
Changed